chainable/__init__.py

```diff
@@ -0,0 +1,46 @@
+from functools import reduce, _initial_missing
+from typing import TypeVar, Callable, overload, Generic
+from collections.abc import Iterable, Iterator
+
+__all__ = ["Chainable"]
+
+T = TypeVar("T")
+S = TypeVar("S")
+
+def flat_map_helper(self: "Chainable[T]", mapper: Callable[[T], Iterable[S]]) -> Iterator[S]:
+    for val in self:
+        yield from mapper(val)
+    
+
+class Chainable(Generic[T]):
+    def __init__(self, it: Iterable[T]) -> None:
+        self.it = it
+
+    def __iter__(self) -> Iterator[T]:
+        return iter(self.it)
+
+    def map(self, mapper: Callable[[T], S]) -> "Chainable[S]":
+        return Chainable(map(mapper, self))
+    
+    def filter(self, predicate: Callable[[T], bool]) -> "Chainable[T]":
+        return Chainable(filter(predicate, self))
+    
+    @overload
+    def reduce(self, reducer: Callable[[T, T], T]) -> T:
+        ...
+    
+    @overload
+    def reduce(self, reducer: Callable[[T, T], T], initial: T) -> T:
+        ...
+
+    def reduce(self, reducer, initial = _initial_missing) -> T:
+        return reduce(reducer, self, _initial_missing)
+
+    def any(self, pred: Callable[[T], bool]) -> bool:
+        return any(pred(x) for x in self)
+    
+    def all(self, pred: Callable[[T], bool]) -> bool:
+        return all(pred(x) for x in self)
+    
+    def flat_map(self, mapper: Callable[[T], Iterable[S]]) -> "Chainable[S]":
+        return Chainable(flat_map_helper(self, mapper))
```

Comment:
The more I think about it, is this method maybe trying to do too much? Why not

```suggestion
    def flatten(self):
        return Chainable(itertools.chain.from_iterable(self))
```

And then let the caller compose `flatten` and `map` as needed?

I'm unsure what's the correct return type to annotate.

Reply:
Well `flat_map` is present in the stdlib of many of the most popular languages:

[JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap)
[C++](https://en.cppreference.com/w/cpp/container/flat_map)
[Rust](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flat_map)
[Java](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#flatMap-java.util.function.Function-)
[category theory](https://en.wikipedia.org/wiki/Kleisli_category)

Not present: C#, Go, Python. I think it's worth including just for how common it is.

Reply:
Maybe so. I still think it's a bad pattern, since `.map(func).flatten()` is less spiraly (`map_and_flatten` would be more accurately named), more composable and extensible, and it keeps all operations to single words. If you want a flat_map, I'm okay with that, but I'd like to see it implemented as the composition of .map and .flatten.

```suggestion
    def flat_map(self, mapper: Callable[[T], Iterable[S]]) -> "Chainable[S]":
        return self.map(mapper).flatten()
```